FirebaseAppDistribution/Sources/FIRFADApiService.m

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#import "FirebaseAppDistribution/Sources/FIRFADApiService.h"
+#import <Foundation/Foundation.h>
+#import "FirebaseAppDistribution/Sources/FIRFADLogger.h"
+#import "FirebaseCore/Sources/Private/FirebaseCoreInternal.h"
+#import "FirebaseInstallations/Source/Library/Private/FirebaseInstallationsInternal.h"
+
+NSString *const kFIRFADApiErrorDomain = @"com.firebase.appdistribution.api";
+NSString *const kFIRFADApiErrorDetailsKey = @"details";
+NSString *const kHTTPGet = @"GET";
+// The App Distribution Tester API endpoint used to retrieve releases
+NSString *const kReleasesEndpointURLTemplate =
+    @"https://firebaseapptesters.googleapis.com/v1alpha/devices/"
+    @"-/testerApps/%@/installations/%@/releases";
+NSString *const kInstallationAuthHeader = @"X-Goog-Firebase-Installations-Auth";
+NSString *const kApiHeaderKey = @"X-Goog-Api-Key";
+NSString *const kResponseReleasesKey = @"releases";
+
+@implementation FIRFADApiService
+
++ (void)generateAuthTokenWithCompletion:(FIRFADGenerateAuthTokenCompletion)completion {
+  FIRInstallations *installations = [FIRInstallations installations];
+
+  // Get a FIS Authentication Token.
+  [installations authTokenWithCompletion:^(
+                     FIRInstallationsAuthTokenResult *_Nullable authTokenResult,
+                     NSError *_Nullable error) {
+    if ([self handleError:&error
+              description:@"Failed to generate Firebase Installation Auth Token."
+                     code:FIRFADApiTokenGenerationFailure]) {
+      FIRFADErrorLog(@"Error getting fresh auth tokens. Error: %@", [error localizedDescription]);
+
+      completion(nil, nil, error);
+      return;
+    }
+
+    [installations installationIDWithCompletion:^(NSString *__nullable identifier,
+                                                  NSError *__nullable error) {
+      if ([self handleError:&error
```

Comment:
Same comment here re: this is run on main thread. Might be totally fine though, but for our situation we needed to put the block on a background thread.

Reply:
Didn't realize these were nested calls, this is probably fine